core/src/main/scala/org/apache/spark/sql/execution/oplog/impl/OpLogRdd.scala

```diff
@@ -231,16 +231,88 @@ class OpLogRdd(
         val data = new SerializedArray(8)
         if (array != null) {
           data.pointTo(array, Platform.BYTE_ARRAY_OFFSET, array.length)
-          data.toArray(arrayType.elementType)
+          if (arrayType.elementType == DateType) {
+            data.toArray[Integer](IntegerType).map(daysSinceEpoch => {
+              if (daysSinceEpoch == null) null else DateTimeUtils.toJavaDate(daysSinceEpoch)
+            })
+          } else if (arrayType.elementType == TimestampType) {
+            data.toArray[Object](LongType).map(nsSinceEpoch => {
+              if (nsSinceEpoch == null) null
+              else new Timestamp(nsSinceEpoch.asInstanceOf[Long] / 1000)
+            })
+          } else {
+            data.toArray(arrayType.elementType)
+          }
         } else null
       case mapType: MapType =>
         val map = valueArr(complexFieldIndex)
         val data = new SerializedMap()
+        // todo: refactor following code.
         if (map != null) {
           data.pointTo(map, Platform.BYTE_ARRAY_OFFSET)
-          val jmap = new java.util.HashMap[Any, Any](data.numElements())
-          data.foreach(mapType.keyType, mapType.valueType, (k, v) => jmap.put(k, v))
-          jmap
+          val jMap = new java.util.HashMap[Any, Any](data.numElements())
```

Comment:
Can refactor this part of code.

Reply:
Will take up in the next iteration. Will keep this PR open until then.
Thanks.